jitbuilder/lljb/src/IRVisitor.cpp

```diff
@@ -0,0 +1,559 @@
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <cassert>
+#include <vector>
+
+namespace lljb
+{
+
+void
+IRVisitor::visitInstruction(llvm::Instruction &I)
+   {
+   llvm::outs() << "Unimplemented instruction being visited: " << I << "\n";
+   assert(0 && "Unimplemented instruction!");
+   }
+
+void
+IRVisitor::visitReturnInst(llvm::ReturnInst &I)
+   {
+   if (!I.getNumOperands())
+      {
+      _builder->Return();
+      }
+   else
+      {
+      llvm::Value * value = I.getOperand(0);
+      TR::IlValue * ilValue = getIlValue(value);
+      _builder->Return(ilValue);
+      }
+   }
+
+void
+IRVisitor::visitAllocaInst(llvm::AllocaInst &I)
+   {
+   TR::IlValue * ilValue = nullptr;
+   if (I.getAllocatedType()->isStructTy())
+      {
+      llvm::StructType * llvmStruct = llvm::dyn_cast<llvm::StructType>(I.getAllocatedType());
+      ilValue = _builder->CreateLocalStruct(_methodBuilder->getIlType(_td,llvmStruct));
+      unsigned elIndex = 0;
+      for (auto elIter = llvmStruct->element_begin();
+            elIter != llvmStruct->element_end(); ++elIter)
+         {
+         llvm::Type * fieldType = llvmStruct->getElementType(elIndex);
+         if (fieldType->isArrayTy())
+            {
+            TR::IlValue * arrayField = _builder->CreateLocalArray(
+               fieldType->getArrayNumElements(),
+               _methodBuilder->getIlType(_td,fieldType));
+            _builder->StoreAt(
+               _builder->StructFieldInstanceAddress(
+                  llvmStruct->getStructName().data(),
+                  _methodBuilder->getMemberNameFromIndex(elIndex),
+                  ilValue),
+               arrayField);
+            }
+         elIndex++;
+         }
+      }
+   else if (I.getAllocatedType()->isArrayTy())
+      {
+      llvm::Type * type = I.getAllocatedType();
+      unsigned numElements = type->getArrayNumElements();
+      type = type->getArrayElementType();
+      while (1)
+         {
+         if (type->isArrayTy())
+            {
+            numElements *= type->getArrayNumElements();
+            type = type->getArrayElementType();
+            }
+         else
+            {
+            break;
+            }
+         }
+      ilValue = _builder->CreateLocalArray(
+                           numElements,
+                           _methodBuilder->getIlType(_td,type));
+      }
+   else
+      {
+      _methodBuilder->allocateLocal(
+      &I,
+      _methodBuilder->getIlType(_td, I.getAllocatedType()));
+      }
+   _methodBuilder->mapIRtoIlValue(&I, ilValue);
+   }
+
+void
+IRVisitor::visitLoadInst(llvm::LoadInst &I)
+   {
+   llvm::Value * source = I.getPointerOperand();
+   TR::IlValue * loadedVal = nullptr;
+   if (_methodBuilder->isIndirectLoadOrStore(source))
+      {
+      TR::IlValue * ilSrc = getIlValue(source);
+      loadedVal = _builder->LoadAt(
+                              _td->PointerTo(_methodBuilder->getIlType(_td,I.getType())),
+                              _builder->
```

Comment:
does this `IndexAt` do anything? I would have expected this whole `LoadAt` could just take i1Src as its operand.

Reply:
You are right. Using `IndexAt` was an extra step required during the early days of the project where local variables were being allocated as single-element arrays and the `ilSrc` was the `IlValue` representing the array. I got rid of this as it is now redundant.